examples/src/main/scala/com/gu/tableversions/examples/DatePartitionedTableLoader.scala

```diff
@@ -1,47 +1,56 @@
 package com.gu.tableversions.examples
 
-import java.net.URI
 import java.sql.{Date, Timestamp}
 
+import cats.effect.IO
+import com.gu.tableversions.core.TableVersions.UserId
+import com.gu.tableversions.core._
 import com.gu.tableversions.examples.DatePartitionedTableLoader.Pageview
-import org.apache.spark.sql.{Dataset, SaveMode, SparkSession}
+import com.gu.tableversions.metastore.Metastore
+import com.gu.tableversions.spark.VersionedDataset
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.sql.{Dataset, SparkSession}
 
 /**
   * This example contains code that writes example event data to a table that has a single date partition.
   * It demonstrates how individual partitions in such a table can be updated using versioning.
-  *
-  * @param tableName the fully qualified table name that will be populated by this loader
-  * @param tableLocation The location where the table data will be stored
   */
-class DatePartitionedTableLoader(tableName: String, tableLocation: URI)(implicit val spark: SparkSession) {
+class DatePartitionedTableLoader(table: TableDefinition)(
+    implicit tableVersions: TableVersions[IO],
+    metastore: Metastore[IO],
+    spark: SparkSession)
+    extends LazyLogging {
 
   import spark.implicits._
 
   def initTable(): Unit = {
-    val ddl = s"""CREATE EXTERNAL TABLE IF NOT EXISTS $tableName (
+    // Create table schema in metastore
+    val ddl = s"""CREATE EXTERNAL TABLE IF NOT EXISTS ${table.name.fullyQualifiedName} (
                  |  `id` string,
                  |  `path` string,
                  |  `timestamp` timestamp
                  |)
                  |PARTITIONED BY (`date` date)
                  |STORED AS parquet
-                 |LOCATION '$tableLocation'
+                 |LOCATION '${table.location}'
     """.stripMargin
 
     spark.sql(ddl)
-    ()
+
+    // Initialise version tracking for table
+    tableVersions.init(table.name).unsafeRunSync()
   }
 
   def pageviews(): Dataset[Pageview] =
-    spark.table(tableName).as[Pageview]
-
-  def insert(dataset: Dataset[Pageview]): Unit = {
-    // Currently, this just uses the basic implementation of writing data to tables via Hive.
-    // This will not do any versioning as-is - this is the implementation we need to replace
-    // with new functionality in this project.
-    dataset.write
-      .mode(SaveMode.Overwrite)
-      .insertInto(tableName)
+    spark.table(table.name.fullyQualifiedName).as[Pageview]
+
+  def insert(dataset: Dataset[Pageview], userId: UserId, message: String): Unit = {
+    import VersionedDataset._
+
+    val (latestVersion, metastoreChanges) = dataset.versionedInsertInto(table, userId, message)
```

Comment:
This is the key bit of this PR, in that this is the code that uses the new API for inserting a dataset into a table in a versioned way as opposed to the previous "insert via Hive" way.